File: contracts/api_types.py
"""
Cricket App API Types

Python dataclasses that EXACTLY mirror the TypeScript types in api_types.ts.
THESE TYPES ARE LAW - all implementations must conform exactly.

Version: 1.0.0
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Literal, Optional, Union


# =============================================================================
# ENUMS (Literal types in TypeScript)
# =============================================================================

class ShotOutcome(str, Enum):
    DOT = "dot"
    ONE = "1"
    TWO = "2"
    THREE = "3"
    FOUR = "4"
    SIX = "6"
    CAUGHT = "caught"
    DROPPED = "dropped"
    MISFIELD = "misfield"


class BallResult(str, Enum):
    DOT = "dot"
    ONE = "1"
    TWO = "2"
    THREE = "3"
    FOUR = "4"
    SIX = "6"
    WICKET = "W"
    WIDE = "wd"
    NO_BALL = "nb"


class Difficulty(str, Enum):
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


class BattingHand(str, Enum):
    RIGHT = "right"
    LEFT = "left"


class ConnectionState(str, Enum):
    DISCONNECTED = "disconnected"
    CONNECTING = "connecting"
    CONNECTED = "connected"
    RECONNECTING = "reconnecting"


class SessionState(str, Enum):
    NONE = "none"
    ACTIVE = "active"
    ENDED = "ended"


class TrackingState(str, Enum):
    IDLE = "idle"
    DETECTING = "detecting"
    TRACKING = "tracking"
    COMPLETED = "completed"


class CatchType(str, Enum):
    REGULATION = "regulation"
    HARD = "hard"
    SPECTACULAR = "spectacular"


# =============================================================================
# DATA STRUCTURES
# =============================================================================

@dataclass(frozen=True)
class Position:
    """Position in field coordinates (metres from batter)."""
    x: float  # metres from batter
    y: float  # metres from batter (positive = toward bowler)


@dataclass(frozen=True)
class FielderConfig:
    """Fielder position configuration.

    Note: id is optional - UI can generate for drag tracking,
    but the game engine doesn't require it.
    """
    x: float  # screen X position (0-100%)
    y: float  # screen Y position (0-100%)
    name: str  # zone name e.g. 'cover', 'mid-off'
    id: Optional[str] = None


@dataclass(frozen=True)
class TrajectoryData:
    """Complete ball trajectory data."""
    projected_distance: float  # total distance in metres (aerial + rolling)
    aerial_distance: float     # distance traveled through air
    rolling_distance: float    # distance ball rolls after landing
    max_height: float          # peak height in metres
    landing_x: float           # X landing coordinate
    landing_y: float           # Y landing coordinate
    final_x: float             # X where ball stops
    final_y: float             # Y where ball stops
    time_of_flight: float      # seconds in air
    horizontal_speed: float    # m/s along ground
    vertical_speed: float      # m/s initial vertical
    direction_x: float         # unit vector X component
    direction_y: float         # unit vector Y component


@dataclass(frozen=True)
class CatchAnalysis:
    """Detailed catch difficulty breakdown.

    Note: Python uses snake_case internally, but JSON wire format uses camelCase
    to match TypeScript conventions. Serialization layer must convert:
    - can_catch -> canCatch
    - catch_type -> catchType
    - reaction_time -> reactionTime
    etc.
    """
    can_catch: bool
    difficulty: float          # 0-1, higher = harder
    catch_type: Optional[CatchType]
    reaction_time: float       # seconds available to react
    movement_required: float   # metres fielder must move
    movement_possible: float   # metres fielder can move given time
    ball_speed_at_fielder: float  # km/h
    height_at_intercept: float    # metres
    time_to_intercept: float      # seconds until ball reaches intercept point
    fielder_arrival_time: float   # seconds for fielder to reach intercept point
    arrived_before_landing: bool  # did fielder get there before ball landed?


@dataclass
class SimulationResult:
    """Complete simulation outcome."""
    outcome: ShotOutcome
    runs: int  # 0-6
    is_boundary: bool
    is_aerial: bool
    end_position: Position
    description: str
    fielder_involved: Optional[str] = None
    fielder_position: Optional[Position] = None  # where fielder started
    fielding_position: Optional[Position] = None  # where fielder collected ball
    trajectory: Optional[TrajectoryData] = None
    catch_analysis: Optional[CatchAnalysis] = None
    fielding_time: Optional[float] = None  # total time from bat to stumps
    collection_difficulty: Optional[float] = None  # 0-1, how rushed was fielder
    alignment_score: Optional[float] = None  # 0-1, how direct ball path was
    priority_score: Optional[float] = None  # combined weighted score
    fielder_arrival_time: Optional[float] = None
    ball_arrival_time: Optional[float] = None


@dataclass(frozen=True)
class SlimShotResult:
    """Minimal shot data for real-time WebSocket display.
    Full SimulationResult stored in database - use REST API GET /api/sessions/:id/deliveries for details."""
    outcome: ShotOutcome
    runs: int
    description: str
    end_position: Position  # For wagon wheel
    fielder_involved: Optional[str]


@dataclass(frozen=True)
class RadarTrackingData:
    """Data captured from radar during ball tracking."""
    timestamp: str  # ISO 8601
    exit_speed: float  # km/h
    horizontal_angle: float  # degrees from straight (-180 to 180)
    vertical_angle: float  # degrees from horizontal (0 to 90)
    projected_distance: float
    max_height: float
    landing_x: float
    landing_y: float
    frames_captured: int
    detection_confidence: float  # 0-1


@dataclass
class Over:
    """A single over of deliveries."""
    balls: list[BallResult]
    runs: int


@dataclass
class SessionData:
    """Complete session data."""
    id: str
    player_id: str
    date: str  # ISO 8601
    runs: int
    balls: int
    fours: int
    sixes: int
    wickets: int
    is_out: bool
    overs: list[Over]
    strike_rate: float
    field_config: list[FielderConfig]  # sessions always have a field config


@dataclass(frozen=True)
class Profile:
    """Player profile."""
    id: str
    name: str
    batting_hand: BattingHand


@dataclass(frozen=True)
class ErrorPayload:
    """Error information."""
    code: str  # error code from error_codes.md
    message: str
    recoverable: bool
    details: Optional[dict] = None


@dataclass(frozen=True)
class WagonWheelShot:
    """A shot displayed on the wagon wheel."""
    id: str
    end_x: float  # screen X % (0-100)
    end_y: float  # screen Y % (0-100)
    outcome: BallResult
    distance: float  # metres


# =============================================================================
# CLIENT -> SERVER MESSAGES
# =============================================================================

@dataclass
class BaseClientMessage:
    """Base class for client messages."""
    message_id: str  # UUID v4
    timestamp: str   # ISO 8601


@dataclass
class SetFieldPayload:
    fielders: list[FielderConfig]  # 1-11 fielders
    boundary_distance: float = 70.0  # 50-100


@dataclass
class SetFieldMessage(BaseClientMessage):
    payload: SetFieldPayload
    type: Literal["set_field"] = "set_field"


@dataclass
class SetDifficultyPayload:
    difficulty: Difficulty


@dataclass
class SetDifficultyMessage(BaseClientMessage):
    payload: SetDifficultyPayload
    type: Literal["set_difficulty"] = "set_difficulty"


@dataclass
class SelectProfilePayload:
    profile_id: str


@dataclass
class SelectProfileMessage(BaseClientMessage):
    payload: SelectProfilePayload
    type: Literal["select_profile"] = "select_profile"


@dataclass
class CreateProfilePayload:
    name: str  # 1-100 chars
    batting_hand: BattingHand


@dataclass
class CreateProfileMessage(BaseClientMessage):
    payload: CreateProfilePayload
    type: Literal["create_profile"] = "create_profile"


@dataclass
class ManualInputPayload:
    result: BallResult
    is_boundary: bool = False


@dataclass
class ManualInputMessage(BaseClientMessage):
    payload: ManualInputPayload
    type: Literal["manual_input"] = "manual_input"


@dataclass
class StartSessionPayload:
    profile_id: str
    field_config: Optional[list[FielderConfig]] = None
    difficulty: Optional[Difficulty] = None
    notes: Optional[str] = None  # max 500 chars


@dataclass
class StartSessionMessage(BaseClientMessage):
    payload: StartSessionPayload
    type: Literal["start_session"] = "start_session"


@dataclass
class EndSessionPayload:
    session_id: str
    save_to_database: bool = True


@dataclass
class EndSessionMessage(BaseClientMessage):
    payload: EndSessionPayload
    type: Literal["end_session"] = "end_session"


@dataclass
class UndoPayload:
    session_id: str


@dataclass
class UndoMessage(BaseClientMessage):
    payload: UndoPayload
    type: Literal["undo"] = "undo"


@dataclass
class UpdateProfilePayload:
    profile_id: str
    name: Optional[str] = None  # 1-100 chars
    batting_hand: Optional[BattingHand] = None


@dataclass
class UpdateProfileMessage(BaseClientMessage):
    payload: UpdateProfilePayload
    type: Literal["update_profile"] = "update_profile"


@dataclass
class PingMessage(BaseClientMessage):
    """Heartbeat ping from client to keep connection alive."""
    type: Literal["ping"] = "ping"


ClientMessage = Union[
    SetFieldMessage,
    SetDifficultyMessage,
    SelectProfileMessage,
    CreateProfileMessage,
    UpdateProfileMessage,
    ManualInputMessage,
    StartSessionMessage,
    EndSessionMessage,
    UndoMessage,
    PingMessage,
]


# =============================================================================
# SERVER -> CLIENT MESSAGES
# =============================================================================

@dataclass
class BaseServerMessage:
    """Base class for server messages."""
    message_id: str  # UUID v4
    timestamp: str   # ISO 8601
    in_reply_to: Optional[str] = None  # UUID of client message this replies to


@dataclass
class ShotResultPayload:
    """Shot result payload for WebSocket. Uses SlimShotResult for efficient transfer."""
    session_id: str
    ball_number: int
    simulation: SlimShotResult  # Slim payload for WebSocket. Full data via REST API.
    radar_data: Optional[RadarTrackingData] = None


@dataclass
class ShotResultMessage(BaseServerMessage):
    payload: ShotResultPayload
    type: Literal["shot_result"] = "shot_result"


@dataclass
class SessionStatePayload:
    session: Optional[SessionData]
    profiles: list[Profile]
    difficulty: Difficulty
    field_config: list[FielderConfig]
    boundary_distance: float
    active_profile_id: Optional[str] = None


@dataclass
class SessionStateMessage(BaseServerMessage):
    payload: SessionStatePayload
    type: Literal["session_state"] = "session_state"


@dataclass
class WagonWheelUpdatePayload:
    session_id: str
    shot: WagonWheelShot


@dataclass
class WagonWheelUpdateMessage(BaseServerMessage):
    payload: WagonWheelUpdatePayload
    type: Literal["wagon_wheel_update"] = "wagon_wheel_update"


@dataclass
class BallTrackingPayload:
    tracking_state: TrackingState
    current_position: Optional[Position] = None
    current_speed: Optional[float] = None
    frames_captured: int = 0


@dataclass
class BallTrackingMessage(BaseServerMessage):
    payload: BallTrackingPayload
    type: Literal["ball_tracking"] = "ball_tracking"


@dataclass
class ConnectionStatusPayload:
    connection_state: ConnectionState
    session_state: SessionState
    radar_connected: bool
    server_version: str
    radar_status: Optional[str] = None
    uptime_seconds: Optional[float] = None


@dataclass
class ConnectionStatusMessage(BaseServerMessage):
    payload: ConnectionStatusPayload
    type: Literal["connection_status"] = "connection_status"


@dataclass
class ErrorMessage(BaseServerMessage):
    payload: ErrorPayload
    type: Literal["error"] = "error"


@dataclass
class PongMessage(BaseServerMessage):
    """Heartbeat pong response from server."""
    in_reply_to: str  # Required for pong - UUID of the ping message
    type: Literal["pong"] = "pong"


ServerMessage = Union[
    ShotResultMessage,
    SessionStateMessage,
    WagonWheelUpdateMessage,
    BallTrackingMessage,
    ConnectionStatusMessage,
    ErrorMessage,
    PongMessage,
]

WebSocketMessage = Union[ClientMessage, ServerMessage]


# =============================================================================
# UTILITY FUNCTIONS
# =============================================================================

def generate_message_id() -> str:
    """Generate a UUID v4 for message IDs."""
    return str(uuid.uuid4())


def create_timestamp() -> str:
    """Create an ISO 8601 timestamp."""
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def is_client_message(msg: WebSocketMessage) -> bool:
    """Check if a message is a client message."""
    client_types = {
        "set_field",
        "set_difficulty",
        "select_profile",
        "create_profile",
        "update_profile",
        "manual_input",
        "start_session",
        "end_session",
        "undo",
        "ping",
    }
    return msg.type in client_types


def is_server_message(msg: WebSocketMessage) -> bool:
    """Check if a message is a server message."""
    server_types = {
        "shot_result",
        "session_state",
        "wagon_wheel_update",
        "ball_tracking",
        "connection_status",
        "error",
        "pong",
    }
    return msg.type in server_types

File: contracts/api_types.ts
/**
 * Cricket App API Types
 *
 * TypeScript interfaces for WebSocket message payloads.
 * THESE TYPES ARE LAW - all implementations must conform exactly.
 *
 * @version 1.0.0
 */

// =============================================================================
// PRIMITIVE TYPES
// =============================================================================

export type ShotOutcome =
  | "dot"
  | "1"
  | "2"
  | "3"
  | "4"
  | "6"
  | "caught"
  | "dropped"
  | "misfield";

export type BallResult =
  | "dot"
  | "1"
  | "2"
  | "3"
  | "4"
  | "6"
  | "W"
  | "wd"
  | "nb";

export type Difficulty = "easy" | "medium" | "hard";

export type BattingHand = "right" | "left";

export type ConnectionState =
  | "disconnected"
  | "connecting"
  | "connected"
  | "reconnecting";

export type SessionState = "none" | "active" | "ended";

export type TrackingState = "idle" | "detecting" | "tracking" | "completed";

export type CatchType = "regulation" | "hard" | "spectacular";

// =============================================================================
// DATA STRUCTURES
// =============================================================================

export interface Position {
  x: number; // metres from batter
  y: number; // metres from batter (positive = toward bowler)
}

export interface FielderConfig {
  id?: string; // optional - UI can generate for drag tracking, game engine doesn't require
  x: number; // screen X position (0-100%)
  y: number; // screen Y position (0-100%)
  name: string; // zone name e.g. 'cover', 'mid-off'
}

export interface TrajectoryData {
  projected_distance: number; // total distance in metres (aerial + rolling)
  aerial_distance: number; // distance traveled through air
  rolling_distance: number; // distance ball rolls after landing
  max_height: number; // peak height in metres
  landing_x: number; // X landing coordinate
  landing_y: number; // Y landing coordinate
  final_x: number; // X where ball stops
  final_y: number; // Y where ball stops
  time_of_flight: number; // seconds in air
  horizontal_speed: number; // m/s along ground
  vertical_speed: number; // m/s initial vertical
  direction_x: number; // unit vector X component
  direction_y: number; // unit vector Y component
}

export interface CatchAnalysis {
  canCatch: boolean;
  difficulty: number; // 0-1, higher = harder
  catchType: CatchType | null;
  reactionTime: number; // seconds available to react
  movementRequired: number; // metres fielder must move
  movementPossible: number; // metres fielder can move given time
  ballSpeedAtFielder: number; // km/h
  heightAtIntercept: number; // metres
  timeToIntercept: number; // seconds until ball reaches intercept point
  fielderArrivalTime: number; // seconds for fielder to reach intercept point
  arrivedBeforeLanding: boolean; // did fielder get there before ball landed?
}

export interface SimulationResult {
  outcome: ShotOutcome;
  runs: number; // 0-6
  is_boundary: boolean;
  is_aerial: boolean;
  fielder_involved: string | null;
  fielder_position: Position | null; // where fielder started
  fielding_position: Position | null; // where fielder collected ball
  end_position: Position; // where ball ended up
  description: string;
  trajectory: TrajectoryData | null;
  catch_analysis: CatchAnalysis | null;
  fielding_time: number | null; // total time from bat to stumps
  collection_difficulty: number | null; // 0-1, how rushed was fielder
  alignment_score: number | null; // 0-1, how direct ball path was
  priority_score: number | null; // combined weighted score
  fielder_arrival_time: number | null;
  ball_arrival_time: number | null;
}

// Slim payload for real-time WebSocket display (5 fields only).
// Full SimulationResult is stored in database. Use REST API GET /api/sessions/:id/deliveries for detailed data.
export interface SlimShotResult {
  outcome: ShotOutcome;
  runs: number;
  description: string;
  end_position: Position; // For wagon wheel
  fielder_involved: string | null;
}

export interface RadarTrackingData {
  timestamp: string; // ISO 8601
  exit_speed: number; // km/h
  horizontal_angle: number; // degrees from straight (-180 to 180)
  vertical_angle: number; // degrees from horizontal (0 to 90)
  projected_distance: number;
  max_height: number;
  landing_x: number;
  landing_y: number;
  frames_captured: number;
  detection_confidence: number; // 0-1
}

export interface Over {
  balls: BallResult[];
  runs: number;
}

export interface SessionData {
  id: string;
  player_id: string;
  date: string; // ISO 8601
  runs: number;
  balls: number;
  fours: number;
  sixes: number;
  wickets: number;
  is_out: boolean;
  overs: Over[];
  strike_rate: number;
  field_config: FielderConfig[]; // sessions always have a field config
}

export interface Profile {
  id: string;
  name: string;
  batting_hand: BattingHand;
}

export interface ErrorPayload {
  code: string; // error code from error_codes.md
  message: string;
  details: Record<string, unknown> | null;
  recoverable: boolean;
}

// =============================================================================
// WAGON WHEEL
// =============================================================================

export interface WagonWheelShot {
  id: string;
  end_x: number; // screen X % (0-100)
  end_y: number; // screen Y % (0-100)
  outcome: BallResult;
  distance: number; // metres
}

// =============================================================================
// CLIENT -> SERVER MESSAGES
// =============================================================================

interface BaseClientMessage {
  message_id: string; // UUID v4
  timestamp: string; // ISO 8601
}

export interface SetFieldMessage extends BaseClientMessage {
  type: "set_field";
  payload: {
    fielders: FielderConfig[]; // 1-11 fielders
    boundary_distance?: number; // 50-100, default 70
  };
}

export interface SetDifficultyMessage extends BaseClientMessage {
  type: "set_difficulty";
  payload: {
    difficulty: Difficulty;
  };
}

export interface SelectProfileMessage extends BaseClientMessage {
  type: "select_profile";
  payload: {
    profile_id: string;
  };
}

export interface CreateProfileMessage extends BaseClientMessage {
  type: "create_profile";
  payload: {
    name: string; // 1-100 chars
    batting_hand: BattingHand;
  };
}

export interface ManualInputMessage extends BaseClientMessage {
  type: "manual_input";
  payload: {
    result: BallResult;
    is_boundary?: boolean;
  };
}

export interface StartSessionMessage extends BaseClientMessage {
  type: "start_session";
  payload: {
    profile_id: string;
    field_config?: FielderConfig[];
    difficulty?: Difficulty;
    notes?: string; // max 500 chars
  };
}

export interface EndSessionMessage extends BaseClientMessage {
  type: "end_session";
  payload: {
    session_id: string;
    save_to_database?: boolean; // default true
  };
}

export interface UndoMessage extends BaseClientMessage {
  type: "undo";
  payload: {
    session_id: string;
  };
}

export interface UpdateProfileMessage extends BaseClientMessage {
  type: "update_profile";
  payload: {
    profile_id: string;
    name?: string; // 1-100 chars
    batting_hand?: BattingHand;
  };
}

export interface PingMessage extends BaseClientMessage {
  type: "ping";
}

export interface SimulateShotMessage extends BaseClientMessage {
  type: "simulate_shot";
  payload: {
    exit_speed: number; // km/h
    horizontal_angle: number; // degrees (0-360)
    vertical_angle: number; // degrees (0-90)
    field_config: FielderConfig[];
    boundary_distance?: number; // default 70
    difficulty?: Difficulty; // default 'medium'
  };
}

export type ClientMessage =
  | SetFieldMessage
  | SetDifficultyMessage
  | SelectProfileMessage
  | CreateProfileMessage
  | UpdateProfileMessage
  | ManualInputMessage
  | StartSessionMessage
  | EndSessionMessage
  | UndoMessage
  | SimulateShotMessage
  | PingMessage;

// =============================================================================
// SERVER -> CLIENT MESSAGES
// =============================================================================

interface BaseServerMessage {
  message_id: string; // UUID v4
  timestamp: string; // ISO 8601
  in_reply_to?: string | null; // UUID of client message this replies to
}

export interface ShotResultMessage extends BaseServerMessage {
  type: "shot_result";
  payload: {
    session_id: string;
    ball_number: number;
    radar_data: RadarTrackingData | null;
    simulation: SlimShotResult; // Slim payload for WebSocket. Full data via REST API.
  };
}

export interface SessionStateMessage extends BaseServerMessage {
  type: "session_state";
  payload: {
    session: SessionData | null;
    profiles: Profile[];
    active_profile_id: string | null;
    difficulty: Difficulty;
    field_config: FielderConfig[];
    boundary_distance: number;
  };
}

export interface WagonWheelUpdateMessage extends BaseServerMessage {
  type: "wagon_wheel_update";
  payload: {
    session_id: string;
    shot: WagonWheelShot;
  };
}

export interface BallTrackingMessage extends BaseServerMessage {
  type: "ball_tracking";
  payload: {
    tracking_state: TrackingState;
    current_position?: Position | null;
    current_speed?: number | null;
    frames_captured?: number;
  };
}

export interface ConnectionStatusMessage extends BaseServerMessage {
  type: "connection_status";
  payload: {
    connection_state: ConnectionState;
    session_state: SessionState;
    radar_connected: boolean;
    radar_status?: string | null;
    server_version: string;
    uptime_seconds?: number;
  };
}

export interface ErrorMessage extends BaseServerMessage {
  type: "error";
  payload: ErrorPayload;
}

export interface PongMessage extends BaseServerMessage {
  type: "pong";
  in_reply_to: string; // UUID of the ping message (required for pong)
}

export interface SimulateResultMessage extends BaseServerMessage {
  type: "simulate_result";
  payload: {
    simulation: SimulationResult;
  };
}

export type ServerMessage =
  | ShotResultMessage
  | SessionStateMessage
  | WagonWheelUpdateMessage
  | BallTrackingMessage
  | ConnectionStatusMessage
  | ErrorMessage
  | SimulateResultMessage
  | PongMessage;

// =============================================================================
// UTILITY TYPES
// =============================================================================

/** Union of all message types for runtime type guards */
export type WebSocketMessage = ClientMessage | ServerMessage;

/** Extract message type from a message */
export type MessageType<T extends WebSocketMessage> = T["type"];

/** Type guard helpers */
export function isClientMessage(msg: WebSocketMessage): msg is ClientMessage {
  const clientTypes = [
    "set_field",
    "set_difficulty",
    "select_profile",
    "create_profile",
    "update_profile",
    "manual_input",
    "start_session",
    "end_session",
    "undo",
    "simulate_shot",
    "ping",
  ];
  return clientTypes.includes(msg.type);
}

export function isServerMessage(msg: WebSocketMessage): msg is ServerMessage {
  const serverTypes = [
    "shot_result",
    "session_state",
    "wagon_wheel_update",
    "ball_tracking",
    "connection_status",
    "error",
    "simulate_result",
    "pong",
  ];
  return serverTypes.includes(msg.type);
}

/** Generate a UUID v4 */
export function generateMessageId(): string {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

/** Create ISO 8601 timestamp */
export function createTimestamp(): string {
  return new Date().toISOString();
}
